mm2src/coins/utxo/rpc_clients.rs

```diff
@@ -1124,26 +1101,61 @@ impl Drop for ElectrumConnection {
 }
 
 #[derive(Debug)]
-pub struct RpcRequestState<T> {
-    is_request_running: Arc<AtomicBool>,
-    response_subs: Vec<oneshot::Sender<Result<T, JsonRpcError>>>,
+struct ConcurrentRequestState<V> {
+    is_running: bool,
+    subscribers: Vec<RpcReqSub<V>>,
+}
+
+impl<V> ConcurrentRequestState<V> {
+    fn new() -> Self {
+        ConcurrentRequestState {
+            is_running: false,
+            subscribers: Vec::new(),
+        }
+    }
 }
 
 #[derive(Debug)]
-pub struct RpcRequestWrapper<Key, Response> {
-    inner: HashMap<Key, RpcRequestState<Response>>,
+pub struct ConcurrentRequestMap<K, V> {
+    inner: AsyncMutex<HashMap<K, ConcurrentRequestState<V>>>,
 }
 
-impl<Key, Response> RpcRequestWrapper<Key, Response> {
-    fn new() -> Self { RpcRequestWrapper { inner: HashMap::new() } }
+impl<K, V> Default for ConcurrentRequestMap<K, V> {
+    fn default() -> Self {
+        ConcurrentRequestMap {
+            inner: AsyncMutex::new(HashMap::new()),
+        }
+    }
+}
+
+impl<K: Clone + Eq + std::hash::Hash, V: Clone> ConcurrentRequestMap<K, V> {
+    pub fn new() -> ConcurrentRequestMap<K, V> { ConcurrentRequestMap::default() }
 
-    #[allow(dead_code)]
-    async fn wrap_request(
-        &mut self,
-        _key: Key,
-        _request: impl Future<Item = Response, Error = JsonRpcError>,
-    ) -> Result<Response, JsonRpcError> {
-        unimplemented!()
+    async fn wrap_request(&self, request_arg: K, request_fut: RpcRes<V>) -> Result<V, JsonRpcError> {
+        let mut map = self.inner.lock().await;
+        let state = map
+            .entry(request_arg.clone())
+            .or_insert_with(ConcurrentRequestState::new);
+        if state.is_running {
+            let (tx, rx) = async_oneshot::channel();
+            state.subscribers.push(tx);
+            // drop here to avoid holding the lock during await
+            drop(map);
+            rx.await.unwrap()
+        } else {
+            // drop here to avoid holding the lock during await
+            drop(map);
+            let request_res = request_fut.compat().await;
+            let mut map = self.inner.lock().await;
+            let state = map.get_mut(&request_arg).unwrap();
+            for sub in state.subscribers.drain(..) {
+                if sub.send(request_res.clone()).is_err() {
+                    log!("subscriber is dropped");
```

Comment:
What about using `common::log::warn` instead?

Reply:
Right, let me change it.